Add Experimental Sendspin over Cast Support

maximmaxim345 opened the pull request:
## Summary

Adds experimental support for using the Sendspin protocol with Chromecast devices via the [sendspin cast](https://sendspin.github.io/cast/sender.html) receiver app. This enables:

- **Cross-platform sync groups**: Group Chromecast devices with other Sendspin-compatible players (like the Home Assistant Voice Preview Edition, the future web player, or any other Sendspin compatible speakers) for synchronized multi-room audio
- **Flexible device combinations**: Group Chromecast Audio, Google Home, and Cast-enabled TVs with non-Google directly in Music Assistant, without going through the Google Home App

## How to use

⚠️ Sendspin over Cast is still highly experimental so expect bugs.

When "Sendspin mode" is enabled for a Chromecast player:
1. Power on a Chromecast device
2. This will create a new Sendspin player called `Speaker Name (Sendspin)`
3. You can group any Sendspin Speaker when grouping them through the `Speaker Name (Sendspin)` speaker

## Configuration

- **Enable Sendspin mode**: Per-player toggle (requires Sendspin provider enabled)
- **Sync delay (ms)**: Adjust timing offset (positive = delay, negative = advance)

maximmaxim345 committed:
WIP: use `aioresonate` branch with new spec

maximmaxim345 committed:
Update Resonate provider for aioresonate API changes

- Convert metadata times to milliseconds (track_duration, track_progress)
- Update playback_speed from 1 to 1000
- Add ArtworkSource parameter to set_media_art()
- Remove deprecated MediaCommand.SEEK

maximmaxim345 committed:
Add artist artwork

maximmaxim345 committed:
Enable volume and mute support

maximmaxim345 committed:
Fix track progress reporting by using the queues elapsed time

maximmaxim345 committed:
Specify supported group commands

maximmaxim345 committed:
Remove unused volume and mute handling

This was already handled by `aioresonate` and therefore unused.

maximmaxim345 committed:
Also update `disconnect_behavior` on grouping changes from `aioresonate`

maximmaxim345 committed:
Update supported group commands when a new group gets created

maximmaxim345 committed:
Support remaining group commands

maximmaxim345 committed:
Adjust callback signature

maximmaxim345 committed:
Also update `disconnect_behaviour` when group changes come externally

maximmaxim345 committed:
Correctly synchronize grouping changes from `aioresonate`

maximmaxim345 committed:
Merge branch 'dev' into feat/resonate-spec-update

maximmaxim345 committed:
feat(chromecast): add experimental Resonate mode for synchronized streaming

- Add Resonate mode toggle to Chromecast player settings
- Add configurable sync delay option (-1000ms to +1000ms, default -350ms)
- Sync player state (queue, playback, volume, mute) from Resonate player
- Send player ID, name, and sync delay to Cast receiver via custom namespace
- Support runtime sync delay updates without app restart

maximmaxim345 committed:
Merge branch 'dev' into feat/resonate-spec-update

marcelveldt commented:
> Gets very slowly out of sync over time. Stopping and starting playback seams to fix this.
This most probably means that sync is not working at all ? So the low level access to the browser's realtime clock I mean.
Did you test a long session with the web client to rule out an implementation error in the javascript client
?

maximmaxim345 commented:
> This most probably means that sync is not working at all ? So the low level access to the browser's realtime clock I mean.
> Did you test a long session with the web client to rule out an implementation error in the javascript client?

I think its very likely that there is a bug in the js client somewhere.
I noticed a faint echo after playing a couple of hours to 5 js clients but they werent all synced from the start so I may am just imagining it 🤷

Edit: Time syncing was broken in `resonate-js`.
Fixed that so it now keeps playing in sync on Chromecast, Chromecast Audio, iOS Safari, and Android Firefox after playing for 2 hours!

maximmaxim345 committed:
feat(chromecast): adjust default Resonate time correction offset

Change default value from -350ms to -300ms for better out-of-the-box
synchronization after cast app update.

maximmaxim345 committed:
Merge branch 'dev' into feat/chromecast-resonate

Resolve conflicts:
- Keep sendspin provider code in sendspin files
- Keep resonate provider manifest.json
- Keep both aioresonate and aiosendspin in requirements

maximmaxim345 committed:
refactor(chromecast): rename resonate to sendspin

- Rename all resonate references to sendspin in chromecast provider
- Update constants, config entries, and method names
- Remove resonate provider manifest.json
- Remove aioresonate from requirements_all.txt

marcelveldt reviewed (approved):
Let's give this to the beta testers for a good test drive!
